src/QMCWaveFunctions/BsplineFactory/SplineR2R.cpp

```diff
@@ -120,21 +121,12 @@ void SplineR2R<ST>::applyRotation(const ValueMatrix& rot_mat, bool use_stored_co
     std::copy_n(spl_coefs, coefs_tot_size, coef_copy_->begin());
   }
 
-  // Apply rotation the dumb way b/c I can't get BLAS::gemm to work...
-  for (auto i = 0; i < BasisSetSize; i++)
-  {
+  std::vector<ST> rot_mat_padded(Nsplines * Nsplines, 0);
+  for (auto i = 0; i < OrbitalSetSize; i++)
     for (auto j = 0; j < OrbitalSetSize; j++)
-    {
-      const auto cur_elem = Nsplines * i + j;
-      auto newval{0.};
-      for (auto k = 0; k < OrbitalSetSize; k++)
-      {
-        const auto index = i * Nsplines + k;
-        newval += (*coef_copy_)[index] * rot_mat[k][j];
-      }
-      spl_coefs[cur_elem] = newval;
-    }
-  }
+       rot_mat_padded[i * Nsplines + j] = rot_mat.data()[i * OrbitalSetSize + j];
+  BLAS::gemm('N', 'N', Nsplines, BasisSetSize, Nsplines, ST(1.0), rot_mat_padded.data(), Nsplines, (*coef_copy_).data(), Nsplines, ST(0.0), spl_coefs, Nsplines);
```

Comment:
I feel we don't need `rot_mat_padded`.
```
BLAS::gemm('N', 'N', OrbitalSetSize, BasisSetSize, OrbitalSetSize, ST(1.0), rot_mat.data(), OrbitalSetSize, (*coef_copy_).data(), Nsplines, ST(0.0), spl_coefs, Nsplines);
```

Reply:
There's still a precision issue, unless BLAS::gemm can handle matrices with different precision types.

Reply:
I was having trouble getting the BLAS call to work without padding. I am trying to figure it out now

I agree there is still a precision issue. I wasn't sure the best way to handle that either. the spline coeffs are ST type and the rotation is RealType. The BLAS::gemm doesn't like mixed precision types, so I had to do some conversion somewhere. If I changed coef_copy_ back to RealType, and the rotation is RealType, the problem then lies in spl_coefs. It is ST*, and I can't just cast it to RealType* in the case of ST=float. 
Making another temporary copy would be costly memory wise

Reply:
Are you intending to work with single or double precision splines?  I ran into some problems with rotation and single precision splines early on, and since then have only used double precision splines.
The code might need to be split differently based spline precision - call gemm for double precision, and an explicit loop for single precision.   That would at least speed up the double precision case.

The precision issue also relates to the history list vs. global rotation method.  With the history list, the rotations get accumulated in the existing coefficient storage.  Single precision coefficients fail at this accumulation of small changes.  With the global rotation method this might not be an issue, since this accumulation of rotations happens in the rotation matrix.  It might work to compute the rotation matrix in double precision, then convert to single precision, and then do the matrix multiplication.  But that would need to be tested, since the matrix multiplication step might also be sensitive to precision.

Reply:
I would prefer double only, but some of the unit tests use float so I was trying to get it to work in general

Reply:
I would be okay with different code paths for float vs double (use the loop for float and gemm for double).  Using "if constexpr" to test the type should make it straightforward.

Reply:
@markdewing, added a split depending on float vs. double. this should avoid the loss in precision.

@ye-luo still trying to figure out how to avoid adding the padded rotation matrix. I thought it should be as simple as fiddling with the LDx parameters, but having trouble getting it to work

Reply:
A code path split by precision should also avoid the need for the extra padded rotation matrix.

Reply:
@ye-luo I think I have the blas call correct now to avoid needing to construct the padded rotation matrix.

Reply:
> A code path split by precision should also avoid the need for the extra padded rotation matrix.

- Do we need single precision spline coefs if there is a problem for rotations (which we ultimately expect to be as common as Jastrow optimization eventually)?
- Wouldn't it be better to modify the unit tests to conform to the production code? 

BLAS-ification is good future-proofing, and probably needed at some level for GPU compatibility anyways.

Reply:
> I feel we don't need `rot_mat_padded`.
> 
> ```
> BLAS::gemm('N', 'N', OrbitalSetSize, BasisSetSize, OrbitalSetSize, ST(1.0), rot_mat.data(), OrbitalSetSize, (*coef_copy_).data(), Nsplines, ST(0.0), spl_coefs, Nsplines);
> ```

I should have read this in detail sooner. I didn't realize you had already told me how to it. Took me a bit, but I stumbled onto doing correctly